bluemira/optimisation/_nlopt/constraint.py

```diff
@@ -0,0 +1,76 @@
+# bluemira is an integrated inter-disciplinary design tool for future fusion
+# reactors. It incorporates several modules, some of which rely on other
+# codes, to carry out a range of typical conceptual fusion reactor design
+# activities.
+#
+# Copyright (C) 2021-2023 M. Coleman, J. Cook, F. Franza, I.A. Maione, S. McIntosh,
+#                         J. Morris, D. Short
+#
+# bluemira is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# bluemira is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+import enum
+from typing import Optional, Tuple, Union
+
+import numpy as np
+
+from bluemira.optimisation._tools import approx_derivative
+from bluemira.optimisation._typing import OptimiserCallable
+
+
+class ConstraintType(enum.Enum):
+    """Enumeration of constraint types."""
+
+    EQUALITY = enum.auto()
+    INEQUALITY = enum.auto()
+
+
+class Constraint:
```

Comment:
This could be premature optimisation (pun unintended :smile:), but this class and `NloptObjectiveFunction` in `objective.py` seem very similar, could they have a common base or is the more trouble than its worth?

Reply:
You're right, the gradient stuff, at the very least, should be refactored into something common.